csrc/scheduler/mma_utils.cpp

```diff
@@ -1530,6 +1627,72 @@ std::unordered_map<ValGroup, MatmulDomain> MatmulPattern::getDimRoles(
   // If there are other patterns, for example a ValGroup present in only A, then
   // we should raise an exception here.
 
+  std::unordered_map<ValGroup, MatmulDomain> dim_roles;
+
+  if (output->definition()->isA<MatmulOp>()) {
+    // Special case for MatmulOp
+    // torch.matmul has a single M, N, and K dimension and 0 or more batch
+    // dimensions.
+    dim_roles[exact_graph.toGroup(A->axis(-1))] = MatmulDomain::K;
+    NVF_ERROR(A->nDims() > 0 && B->nDims() > 0);
+    size_t m_and_k_dims = 0;
+    if (A->nDims() == 1 && B->nDims() == 1) {
+      NVF_ERROR(
+          false, "MatmulOp node should not be created when both inputs are 1D");
+    } else if (A->nDims() == 1) {
+      // Missing M dimension
+      dim_roles[exact_graph.toGroup(B->axis(-1))] = MatmulDomain::N;
+      m_and_k_dims = 1;
+    } else if (B->nDims() == 1) {
+      // Missing N dimension
+      dim_roles[exact_graph.toGroup(A->axis(-2))] = MatmulDomain::M;
+      m_and_k_dims = 1;
```

Comment:
That's a nice idea. In trying to do this I actually found a logic error in the current approach:

The vector returned by `mapMatmulOpIterDomains` contains `nullptr` when that operand has no dimension to match the output dim. However, that is not quite enough information to determine the role without duplicating the logic above. For example:
```
A [ iS0{b} iS1{M} iS2{K} ]
B [ iS3{K} iS4{N} ]
output [ iS5{b} iS6{M} iS7{N} ]
```
In this case when we get these mappings:
```
mapping_a = [ iS0{b} iS1{M} nullptr ]
mapping_b = [ nullptr nullptr iS4{N} ]
```
In the current approach, `iS5{b}` will be treated as a batch dimension. However, since it is actually being broadcasted in the `B` operand, it is actually another M dimension. Anyway, I think using this approach is nice and simple, and even better, correct.

Reply:
I see. My suggestion was that the role information is embedded in the position of the iterdomain in the mapping output. For eg: `out_size-3, out_size-2, out_size-1` are M, N, K respectively.

Could you clarify which approach is erroneous -- `mappingMatmulOpIterDomain` or this PR? Why should `iS5` be M instead of batch?

Reply:
As for why `iS5` should not be considered a batch dim. For a generic matmul problem, we have two operands and one output and we define the dimension roles as follows:
- The output has one reduction dimension which we call K. It must be non-broadcast in both operands (or else this is not a matmul).
- There's a dimension that is broadcast in the first operand and not the second. It appears in the output (the broadcast is concretized). This is what we call the M dimension.
- Likewise there's a dimension that is broadcast in the second operand but not the first. This is the N dimension.
- There may be dimensions that are Iteration in both operands as well as in the output. This is a batch dimension.

I think the confusion is coming from the difference between a batch matrix multiply, in which we are given a batch of _independent_ matmul problems, versus batched DL training in which we have a batch of data that forms a matrix and we want to comput a batch of mat-vec problems as a single matmul. In the former case the "batch" dimension is really the M dimension of the first argument (input) while the matrix being multiplied against each row of the input is the same (the weights) and it's not batched.